test/test_block_shuffle.cu

```diff
@@ -0,0 +1,358 @@
+{ };
+
+template<typename data_type,
+         unsigned int threads_in_block>
+struct RotateTest : public RotateTestBase<data_type, threads_in_block, 24 /* offset */>
+{ };
+
+
+template <typename data_type, typename test_type>
+int CheckResult(
+        int num_items,
+        const data_type *d_output,
+        data_type *h_output,
+        const test_type &test)
+{
+    CubDebugExit(cudaMemcpy(h_output, d_output, num_items * sizeof (data_type), cudaMemcpyDeviceToHost));
+
+    for (int i = 0; i < num_items; i++)
+    {
+        if (!test.check (h_output, i))
+        {
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+template <
+        typename data_type,
+        unsigned int items_per_thread,
+        unsigned int threads_in_block,
+        template<typename, unsigned int, unsigned int> class test_type>
+void Test(unsigned int num_items,
+          data_type *d_data,
+          data_type *h_data)
+{
+    test_type<data_type, items_per_thread, threads_in_block> test;
+
+    Iota(num_items, d_data);
+    BlockShuffleTest<data_type, items_per_thread, threads_in_block>(d_data, test);
+    AssertEquals(0, CheckResult(num_items, d_data, h_data, test));
+}
+
+template <
+        typename data_type,
+        unsigned int items_per_thread,
+        unsigned int threads_in_block,
+        template<typename, unsigned int> class test_type>
+struct SingleItemTestHelper
+{
+    static void run(unsigned int /* num_items */,
+                    data_type * /* d_data */,
+                    data_type * /* h_data */)
+    {
+    }
+};
+
+template <
+        typename data_type,
+        unsigned int threads_in_block,
+        template<typename, unsigned int> class test_type>
+struct SingleItemTestHelper<data_type, 1, threads_in_block, test_type>
+{
+    static void run(unsigned int num_items,
+                   data_type *d_data,
+                   data_type *h_data)
+    {
+        test_type<data_type, threads_in_block> test;
+
+        Iota(num_items, d_data);
+        BlockShuffleTest<data_type, 1 /* items_per_thread */, threads_in_block>(d_data, test);
+        AssertEquals(0, CheckResult(num_items, d_data, h_data, test));
+    }
+};
+
+
+template <
+        typename data_type,
+        unsigned int items_per_thread,
+        unsigned int threads_in_block>
+void Test(
+    CachingDeviceAllocator &g_allocator
+)
+{
+    const unsigned int num_items = items_per_thread * threads_in_block;
+
+    data_type *d_data = nullptr;
+    CubDebugExit(g_allocator.DeviceAllocate((void**)&d_data, sizeof(data_type) * num_items));
+
+    std::unique_ptr<data_type[]> h_data(new data_type[num_items]);
+
+    Test<data_type, items_per_thread, threads_in_block, UpTest>(num_items, d_data, h_data.get());
+    Test<data_type, items_per_thread, threads_in_block, DownTest>(num_items, d_data, h_data.get());
+
+    SingleItemTestHelper<data_type, items_per_thread, threads_in_block, OffsetUpTest>().run(num_items,
```

Comment:
Can you add a comment explaining what this helper is used for? It looks like this becomes a no-op when `items_per_thread` != 1, but this is only invoked (in main) with `items_per_thread` == 1.

We should probably add some test cases where `items_per_thread` > 1.